modules/ismp/clients/grandpa/src/benchmarking.rs

```diff
@@ -0,0 +1,84 @@
+#![cfg(feature = "runtime-benchmarks")]
+use super::*;
+use frame_benchmarking::v2::*;
+use frame_support::pallet_prelude::Weight;
+use frame_system::RawOrigin;
+use sp_std::prelude::*;
+
+/// Benchmarks for the ISMP GRANDPA pallet operations
+#[benchmarks]
+mod benchmarks {
+    use super::*;
+
+    /// Benchmark for add_state_machines extrinsic
+    /// The benchmark creates n state machines and measures the time to add them
+    /// to the whitelist.
+    ///
+    /// Parameters:
+    /// - `n`: Number of state machines to add in a single call
+    #[benchmark]
+    fn add_state_machines(n: Linear<1, 100>) -> Result<(), BenchmarkError> {
+        let caller: T::AccountId = whitelisted_caller();
+        
+        let state_machines: Vec<AddStateMachine> = (0..n)
+            .map(|i| {
+                let id = [i as u8, 0, 0, 0]; // Create unique 4-byte identifier
+                AddStateMachine {
+                    state_machine: StateMachine::Substrate(id),
+                    slot_duration: 6000u64,
+                }
+            })
+            .collect();
+
+        #[extrinsic_call]
+        _(RawOrigin::Root, state_machines);
+
+        // Verify operation was successful
+        assert!(SupportedStateMachines::<T>::iter().count() == n as usize);
+        Ok(())
+    }
+
+    /// Benchmark for remove_state_machines extrinsic
+    /// The benchmark first adds n state machines, then measures the time to remove them
+    /// from the whitelist.
+    ///
+    /// Parameters:
+    /// - `n`: Number of state machines to remove in a single call
+    #[benchmark]
+    fn remove_state_machines(n: Linear<1, 100>) -> Result<(), BenchmarkError> {
+        let caller: T::AccountId = whitelisted_caller();
+        
+        // Setup: First add state machines that we'll remove
+        let setup_machines: Vec<AddStateMachine> = (0..n)
+            .map(|i| {
+                let id = [i as u8, 0, 0, 0]; // Create unique 4-byte identifier
+                AddStateMachine {
+                    state_machine: StateMachine::Substrate(id),
+                    slot_duration: 6000u64,
+                }
+            })
+            .collect();
+
+        // Add the machines using root origin
+        Pallet::<T>::add_state_machines(
+            RawOrigin::Root.into(),
+            setup_machines.clone(),
+        )?;
+
+        // Create removal list
+        let remove_machines: Vec<StateMachine> = 
+            setup_machines.into_iter().map(|m| m.state_machine).collect();
+
+        // Verify initial state
+        assert!(SupportedStateMachines::<T>::iter().count() == n as usize);
+
+        #[extrinsic_call]
+        _(RawOrigin::Root, remove_machines);
+
+        // Verify all machines were removed
+        assert!(SupportedStateMachines::<T>::iter().count() == 0);
+        Ok(())
+    }
+}
+
+impl_benchmark_test_suite!(Pallet, crate::mock::new_test_ext(), crate::mock::Test);
```

Comment:
This line doesn't compile

Reply:
<img width="611" alt="Screenshot 2024-12-17 at 5 31 53 PM" src="https://github.com/user-attachments/assets/ba2d8ec8-eeba-4547-8151-ca4f708ccbb8" />
Hi @Wizdave97,
I've now integrated direct testing against OpSepolia as requested. The tests run successfully against the OpSepolia endpoint, verifying both the L2 data cost calculations and the gas estimation logic in a live environment. Additionally, I've refined the code comments and references to the OpStack/Ecotone specifications for clarity. Let me know if there's anything else you'd like addressed!

Reply:
This is the wrong pr

Reply:
Sorry for the clusterf*** : /
This PR adds benchmarking support for ismp-grandpa (addressing #350), with empirical weight calculations derived from benchmarking results.

Key Changes:
- Added benchmarking infrastructure (WeightInfo trait, benchmarking module)
- Implemented dynamic weight calculations based on input size
- Integrated with runtime benchmarking system

Benchmark Results:
- add_state_machines: base 5.525 µs + 1.458 µs per item
- remove_state_machines: base 4.914 µs + 1.419 µs per item

To reproduce the benchmarks:
```bash
# Build with benchmarking support
cargo build --release --features runtime-benchmarks

# Run benchmarks
cargo run --release --features runtime-benchmarks --bin hyperbridge -- benchmark pallet \
    --runtime ./target/release/wbuild/gargantua-runtime/gargantua_runtime.compact.wasm \
    --pallet ismp_grandpa \
    --extrinsic '*' \
    --steps 50 \
    --repeat 20 \
    --wasm-execution=compiled \
    --genesis-builder=runtime

[benchmarks.txt](https://github.com/user-attachments/files/18173989/benchmarks.txt)